Visualization/generate_BackBLAST_heatmap.R

```diff
@@ -0,0 +1,618 @@
+  
+}
+
+
+#' Convert character "NA" (from command line) into true NA
+#' 
+#' @param entry single-length vector of any type
+#' @return single-length vector; if it was "NA", it will now be NA; otherwise it will be the same as input
+#' @export
+convert_to_true_NA <- function(entry) {
+  
+  if (is.na(entry) == TRUE) {
+    
+    return(entry)
+    
+  } else if (entry == "NA") {
+    
+    flog.debug("Converting to true NA")
+    return(NA)
+    
+  } else {
+    return(entry)
+  }
+}
+
+
+#' Re-root a ggtree
+#' 
+#' @param phylo_tree ggtree-format tree
+#' @param root_name Character (length 1 vector); the exact name of the to-be root
+#' @return ggtree-format tree, rooted
+#' @export
+reroot_ggtree <- function(phylo_tree, root_name) {
+  
+  # Extract the data from the tree in tabular format
+  tree_data <- ggtree::ggtree(phylo_tree)$data
+  
+  # Check that the root_name exists
+  if ( (root_name %in% tree_data$label) == FALSE ) {
+    futile.logger::flog.error(glue::glue("Could not find the root_name '", root_name, 
+                          "' in the provided tree. Cannot re-root. Exiting..."))
+    quit(status = 1, save = "no")
+  }
+  
+  # Get the ggtree ID # of the to-be root
+  tip_label_index <- match(x = root_name, table = phylo_tree$tip.label)
+  
+  # Check that only one matching parent node exists
+  if (length(tip_label_index) != 1) {
+    futile.logger::flog.error(glue::glue("The provided root_name '", root_name, 
+                          "' appears to have more than one associated node. Cannot re-root. Exiting..."))
+    quit(status = 1, save = "no")
+  }
+  
+  # Re-root
+  tree_rooted <- ggtree::reroot(phylo_tree, node = tip_label_index)
+  
+  return(tree_rooted)
+}
+
+
+#' Extracts tree label (bootstrap) data and applies bootstrap cutoff if desired
+#' 
+#' @param phylo_tree ggtree-format tree
+#' @param bootstrap_cutoff Numeric (length 1 vector); the minimum bootstrap cutoff, in percent
+#' @return extracted tree data as a data frame with labels adjusted to only contain bootstraps (numeric) (optionally) above bootstrap cutoff for branches.
+#' To be mapped onto the main tree during plotting.
+#' @export
+generate_bootstrap_labels <- function(phylo_tree, bootstrap_cutoff) {
+  # For adding bootstrap labels like this, see https://guangchuangyu.github.io/software/ggtree/faq/ (accessed Sept 14, 2018)
+  # Note that 'labels' encompasses tip labels and node labels. You can to extract the tip labels alone.
+  
+  # Extract info from the tree into data frame format
+  bootstrap_label_data <- ggtree::ggtree(phylo_tree)$data
+  
+  # Set the labels to numeric (instead of character; will make non character go 'NA', like perhaps tip labels)
+  # This longer lapply function tries to avoid warnings from coersion while still allowing for transparent error reporting otherwise
+  bootstrap_label_data$label <- unlist(lapply(bootstrap_label_data$label, function(x) {
+    if(suppressWarnings(as.numeric(x)) %>% is.na == FALSE) {
```

Comment:
What is the variable ```x``` please give it a name. Is this the label?

Reply:
Indeed -- name is revised